pkg/controller/pinned-image-set/pinned_image_set_controller.go

```diff
@@ -0,0 +1,698 @@
+}
+
+func (ctrl *Controller) enqueue(cfg *mcfgv1.PinnedImageSet) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(cfg)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %w", cfg, err))
+		return
+	}
+	ctrl.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (ctrl *Controller) worker() {
+	for ctrl.processNextWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextWorkItem() bool {
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	err := ctrl.syncHandler(key.(string))
+	ctrl.handleErr(err, key)
+
+	return true
+}
+
+func (ctrl *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.queue.Forget(key)
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		klog.V(2).Infof("Error syncing PinnedImageSet %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	klog.V(2).Infof("Dropping PinnedImageSet %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+func setPinnedImageSetCondition(conditions *[]metav1.Condition, newCondition metav1.Condition) {
+	if conditions == nil {
+		conditions = &[]metav1.Condition{}
+	}
+	existingCondition := getPinnedImageSetCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		*conditions = append(*conditions, newCondition)
+		return
+	}
+
+	if existingCondition.Status != newCondition.Status {
+		existingCondition.Status = newCondition.Status
+		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
+}
+
+func getPinnedImageSetCondition(conditions []metav1.Condition, conditionType string) *metav1.Condition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *Controller) updateStatus(
+	imageSet *mcfgv1.PinnedImageSet,
+	condition metav1.Condition,
+) error {
+	return retry.RetryOnConflict(ctrl.waitBackOff, func() error {
+		existingImageSet, err := ctrl.imageSetLister.Get(imageSet.Name)
+		if err != nil {
+			return err
+		}
+
+		newImageSet := existingImageSet.DeepCopy()
+		if imageSet.GetGeneration() != imageSet.Status.ObservedGeneration {
+			imageSet.Status.ObservedGeneration = imageSet.GetGeneration()
+		}
+
+		setPinnedImageSetCondition(&newImageSet.Status.Conditions, condition)
+
+		if equality.Semantic.DeepEqual(existingImageSet.Status, newImageSet.Status) {
+			return nil
+		}
+
+		_, err = ctrl.mcfgClient.MachineconfigurationV1().PinnedImageSets().UpdateStatus(context.TODO(), newImageSet, metav1.UpdateOptions{})
```

Comment:
`UpdateStatus` is right here as opposed to patch. Though you might need the rbac for pinnedimagesets/status specifically? I have run into this before where it does not allow me to update status unless I have this role.

Reply:
ok mcc rbac I believe is inclusive but will doublecheck.

```
- apiGroups: ["machineconfiguration.openshift.io"]
  resources: ["*"]
  verbs: ["*"]
```